DFC.ServiceTaxonomy.GraphSync/Activities/AuditSyncIssuesTask.cs

```diff
@@ -0,0 +1,144 @@
+    {
+        public AuditSyncIssuesTask(
+            ISession session,
+            IGraphDatabase graphDatabase,
+            IMergeGraphSyncer mergeGraphSyncer,
+            IStringLocalizer<AuditSyncIssuesTask> localizer,
+            IContentDefinitionManager contentDefinitionManager,
+            IEnumerable<IContentPartGraphSyncer> partSyncers)
+        {
+            _session = session;
+            _graphDatabase = graphDatabase;
+            _mergeGraphSyncer = mergeGraphSyncer;
+            T = localizer;
+            _partSyncers = partSyncers.ToDictionary(x => x.PartName ?? "Eponymous");
+            _syncFailedContentItems = new List<ContentItem>();
+            _contentTypes = contentDefinitionManager
+                .ListTypeDefinitions()
+                .Where(x => x.Parts.Any(p => p.Name == "GraphSyncPart"))
+                .ToDictionary(x => x.Name);
+        }
+
+        private readonly ISession _session;
+        private readonly IGraphDatabase _graphDatabase;
+        private readonly Dictionary<string, IContentPartGraphSyncer> _partSyncers;
+        private readonly IMergeGraphSyncer _mergeGraphSyncer;
+        private IStringLocalizer T { get; }
+
+        public override string Name => nameof(AuditSyncIssuesTask);
+        public override LocalizedString DisplayText => T["Identify sync issues and retry"];
+        public override LocalizedString Category => T["National Careers Service"];
+
+        private readonly List<ContentItem> _syncFailedContentItems;
+        private readonly Dictionary<string, ContentTypeDefinition> _contentTypes;
+
+        public override IEnumerable<Outcome> GetPossibleOutcomes(WorkflowExecutionContext workflowContext, ActivityContext activityContext)
+        {
+#pragma warning disable S3220 // Method calls should not resolve ambiguously to overloads with "params"
+            return Outcomes(T["Done"]);
+#pragma warning restore S3220 // Method calls should not resolve ambiguously to overloads with "params"
+        }
+
+        public override async Task<ActivityExecutionResult> ExecuteAsync(WorkflowExecutionContext workflowContext, ActivityContext activityContext)
+        {
+            var timestamp = DateTime.UtcNow;
+            var log = (await _session.Query<AuditSyncLog>().FirstOrDefaultAsync()) ??
+                      new AuditSyncLog() {LastSynced = DateTime.MinValue};
+
+            var contentItems = await _session
+                //do we only care about the latest published items?
+                .Query<ContentItem, ContentItemIndex>(x =>
+                    x.ContentType.IsIn(_contentTypes.Keys) && x.Latest && x.Published && (x.CreatedUtc >= log.LastSynced ||  x.ModifiedUtc >= log.LastSynced))
+                .ListAsync();
+
+            foreach (var contentItem in contentItems)
+            {
+                await CheckIfContentItemSynced(contentItem);
+            }
+
+            foreach (var contentItem in _syncFailedContentItems)
```

Comment:
If the content item can't sync correctly for some non-transient reason and `SyncToGraph` doesn't throw, this looks like it could loop forever.
We should probably have a max try count for syncing a single content item.

Reply:
Can't see why it would loop forever? It should only try once per content item in the sync failed list, and anything left in there at the end is still failing?

Reply:
What I was thinking was this scenario...
```
foreach (var contentItem in _syncFailedContentItems)
{
  ...
  _syncFailedContentItems.Remove(contentItem);
  ...
  await CheckIfContentItemSynced(contentItem); => _syncFailedContentItems.Add(contentItem);
}
```
Googling as to whether that would result in an infinite loop, it looks like mutating, especially removing items from a collection that is being iterated over with a foreach, might be a bad idea anyway, although it looks like the behavior might have changed at some point.
Would you mind checking..
1) if mutating the collection within the foreach is ok?
2) that it won't loop forever if you remove then add back an item into the _syncFailedContentItems collection